### PR TITLE
Convert AbstractTensor from struct to class

### PR DESCRIPTION
**Pull request by jacobhinkle**
`AbstractTensor` now has many methods and as of #2933 it is subclassed into `TaggedAbstractTensor` which has additional methods. Also, there is no longer just a single vector attribute `domain` but also `info`. Instead of requiring the user to manage those vectors and keep them the same length, this PR makes them `protected` members of a class and uses the existing accessors to adjust them, as suggested in https://github.com/NVIDIA/Fuser/pull/2963#discussion_r1765951724.

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
!build

**Comment by zasdfgbnm** [comment]
Is the build failure related?

**Comment by jacobhinkle** [comment]
> Is the build failure related?

I am not sure yet, but I doubt it. I am also seeing this failure on #2913 which doesn't include this change.

File: csrc/abstract_tensor.h [hunk 1]
```diff
@@ -1000,8 +1005,13 @@ struct TaggedAbstractTensor : AbstractTensorWithInfo<TagSetInfo<Tag>> {
...
   const std::unordered_set<Tag>& getTags(int64_t i) const {
-    i = wrapDim(i, (int64_t)this->info.size());
-    return this->info[i].tags;
+    i = wrapDim(i, (int64_t)this->size());
+    return this->info_[i].tags;
+  }
+
+  std::unordered_set<Tag>& getTags(int64_t i) {
+    i = wrapDim(i, (int64_t)this->size());
+    return this->info_[i].tags;
```
Comment (jacobhinkle) [inline csrc/abstract_tensor.h:1014]: I added this non-const accessor so that we can manually insert and erase tags for a single axis.

**Comment by zasdfgbnm** [comment]
I am surprised that this PR just work, because I remembered that there are more direct uses of `.domain` and some requires none-trivial change.

**Comment by jacobhinkle** [comment]
> I am surprised that this PR just work, because I remembered that there are more direct uses of `.domain` and some requires none-trivial change.

Yeah I thought it was going to be tougher too but the only places `AbstractTensor` is currently used is in matmul utilities and TMA analysis:
- `MmaSwizzler::scheduleMmaOutputAllocation`
- `swizzleSharedMemory`
- `collapse_tma_domain::DomainMerger`

Only `DomainMerger` was tinkering with `.domain` directly. Otherwise, we do make use of `AbstractTensor` in some tests outside of `test_abstract_tensor.cpp` but those are not directly manipulating `.domain` either.

**Review by zasdfgbnm — COMMENTED** [review]
LGTM, just want to see CI green for double confirming.

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
Rebuilding following merging of #2969

**Review by zasdfgbnm — APPROVED** [review]

**Pull request merged**